pkg/snet/arclient/client.go

```diff
@@ -325,11 +410,121 @@ func (c *httpClient) WS(ctx context.Context, dialCh <-chan cipher.PubKey) (<-cha
 		}
 	}(conn, dialCh)
 
-	return addrCh, nil
+	c.stcphAddrCh = addrCh
+
+	return nil
+}
+
+// LocalAddresses contains outbound port and all network addresses of visor.
+type LocalAddresses struct {
+	Port      string   `json:"port"`
+	Addresses []string `json:"addresses"`
 }
 
-func (c *httpClient) Close() error {
-	return c.wsConn.Close(websocket.StatusNormalClosure, "client closed")
+func (c *client) BindSUDPH(ctx context.Context, conn net.Conn, localPort string) (<-chan RemoteVisor, error) {
+	if c.sudphAddrCh == nil {
+		if err := c.initSUDPH(ctx, conn, localPort); err != nil {
+			return nil, err
+		}
+	}
+
+	return c.sudphAddrCh, nil
+}
+
+func (c *client) initSUDPH(ctx context.Context, conn net.Conn, localPort string) error {
+	addresses, err := localAddresses()
+	if err != nil {
+		return err
+	}
+
+	localAddresses := LocalAddresses{
+		Addresses: addresses,
+		Port:      localPort,
+	}
+
+	laData, err := json.Marshal(localAddresses)
+	if err != nil {
+		return err
+	}
+
+	if _, err := conn.Write(laData); err != nil {
+		return err
+	}
+
+	addrCh := make(chan RemoteVisor, addrChSize)
+
+	go func(conn net.Conn, addrCh chan<- RemoteVisor) {
+		defer func() {
+			close(addrCh)
+		}()
+
+		buf := make([]byte, 4096)
+
+		for {
+			n, err := conn.Read(buf)
+			if err != nil {
+				log.Errorf("Failed to read SUDPH message: %v", err)
+				return
+			}
+
+			log.Infof("New SUDPH message: %v", string(buf[:n]))
+
+			var remote RemoteVisor
+			if err := json.Unmarshal(buf[:n], &remote); err != nil {
+				log.Errorf("Failed to read unmarshal message: %v", err)
+				continue
+			}
+
+			addrCh <- remote
+		}
+	}(conn, addrCh)
+
+	go func() {
+		if err := c.keepAliveLoop(ctx, conn); err != nil {
+			log.WithError(err).Errorf("Failed to send keep alive UDP packet to address-resolver")
+		}
+	}()
+
+	c.sudphAddrCh = addrCh
+
+	return nil
+}
+
+func (c *client) Close() error {
+	defer func() {
+		c.stcphConn = nil
+		// TODO(nkryuchkov): uncomment
```

Comment:
Was it forgotten? As well as closing of sudph conn below

Reply:
@Darkren No, I left it for https://github.com/SkycoinProject/skywire-mainnet/issues/409, because I'm going to combine all direct transports in one and I also need to add redialing logic.